File: src/xai/sdk/python/xai_sdk/clients/ai_client.py
from __future__ import annotations

"""
AI Client for XAI SDK

Handles all AI-related operations including:
- Personal AI assistant operations
- Smart contract creation and deployment
- Transaction optimization
- Blockchain and wallet analysis
- Node setup recommendations
"""

from typing import Any, Iterator

from ..exceptions import APIError
from ..http_client import HTTPClient


class AIClient:
    """Client for AI operations on XAI blockchain."""

    def __init__(self, http_client: HTTPClient) -> None:
        """
        Initialize AI Client.

        Args:
            http_client: HTTP client instance
        """
        self.http_client = http_client

    def atomic_swap(
        self,
        from_currency: str,
        to_currency: str,
        amount: float,
        recipient_address: str,
    ) -> dict[str, Any]:
        """
        Execute atomic swap with AI assistance.

        Args:
            from_currency: Source currency (e.g., "XAI", "BTC")
            to_currency: Target currency
            amount: Amount to swap
            recipient_address: Destination address

        Returns:
            Swap execution result

        Raises:
            APIError: If swap fails
        """
        payload = {
            "from_currency": from_currency,
            "to_currency": to_currency,
            "amount": amount,
            "recipient_address": recipient_address,
        }
        return self.http_client.post("/personal-ai/atomic-swap", data=payload)

    def create_contract(
        self,
        contract_type: str,
        parameters: dict[str, Any],
        description: str | None = None,
    ) -> dict[str, Any]:
        """
        Create smart contract with AI assistance.

        Args:
            contract_type: Type of contract (e.g., "token", "nft", "escrow")
            parameters: Contract parameters
            description: Optional description of contract purpose

        Returns:
            Created contract details

        Raises:
            APIError: If contract creation fails
        """
        payload = {
            "contract_type": contract_type,
            "parameters": parameters,
        }
        if description:
            payload["description"] = description
        return self.http_client.post("/personal-ai/smart-contract/create", data=payload)

    def deploy_contract(
        self,
        contract_code: str,
        constructor_args: list[Any] | None = None,
        gas_limit: int | None = None,
    ) -> dict[str, Any]:
        """
        Deploy smart contract with AI optimization.

        Args:
            contract_code: Contract bytecode or source
            constructor_args: Optional constructor arguments
            gas_limit: Optional gas limit

        Returns:
            Deployment result with contract address

        Raises:
            APIError: If deployment fails
        """
        payload = {
            "contract_code": contract_code,
        }
        if constructor_args:
            payload["constructor_args"] = constructor_args
        if gas_limit:
            payload["gas_limit"] = gas_limit
        return self.http_client.post("/personal-ai/smart-contract/deploy", data=payload)

    def optimize_transaction(
        self,
        transaction: dict[str, Any],
        optimization_goals: list[str] | None = None,
    ) -> dict[str, Any]:
        """
        Optimize transaction with AI.

        Args:
            transaction: Transaction to optimize
            optimization_goals: Optional goals like ["low_fee", "fast_confirmation"]

        Returns:
            Optimized transaction details

        Raises:
            APIError: If optimization fails
        """
        payload = {
            "transaction": transaction,
        }
        if optimization_goals:
            payload["optimization_goals"] = optimization_goals
        return self.http_client.post("/personal-ai/transaction/optimize", data=payload)

    def analyze_blockchain(
        self,
        query: str,
        context: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """
        Analyze blockchain with AI.

        Args:
            query: Analysis query or question
            context: Optional additional context

        Returns:
            Analysis results

        Raises:
            APIError: If analysis fails
        """
        payload = {"query": query}
        if context:
            payload["context"] = context
        return self.http_client.post("/personal-ai/analyze", data=payload)

    def analyze_wallet(self, address: str) -> dict[str, Any]:
        """
        Analyze wallet with AI.

        Args:
            address: Wallet address to analyze

        Returns:
            Wallet analysis results

        Raises:
            APIError: If analysis fails
        """
        return self.http_client.post("/personal-ai/wallet/analyze", data={"address": address})

    def wallet_recovery_advice(
        self,
        partial_info: dict[str, Any],
    ) -> dict[str, Any]:
        """
        Get wallet recovery advice from AI.

        Args:
            partial_info: Partial wallet information (e.g., mnemonic words, dates)

        Returns:
            Recovery recommendations

        Raises:
            APIError: If recovery advice fails
        """
        return self.http_client.post("/personal-ai/wallet/recovery", data=partial_info)

    def node_setup_recommendations(
        self,
        hardware_specs: dict[str, Any] | None = None,
        use_case: str | None = None,
    ) -> dict[str, Any]:
        """
        Get node setup recommendations from AI.

        Args:
            hardware_specs: Optional hardware specifications
            use_case: Optional use case (e.g., "mining", "validator", "archive")

        Returns:
            Setup recommendations

        Raises:
            APIError: If recommendations fail
        """
        payload: dict[str, Any] = {}
        if hardware_specs:
            payload["hardware_specs"] = hardware_specs
        if use_case:
            payload["use_case"] = use_case
        return self.http_client.post("/personal-ai/node/setup", data=payload)

    def liquidity_alert(
        self,
        pool_id: str,
        alert_type: str,
        threshold: float | None = None,
    ) -> dict[str, Any]:
        """
        Set up liquidity pool alert with AI monitoring.

        Args:
            pool_id: Liquidity pool identifier
            alert_type: Type of alert (e.g., "price_change", "volume", "impermanent_loss")
            threshold: Optional threshold value

        Returns:
            Alert configuration result

        Raises:
            APIError: If alert setup fails
        """
        payload = {
            "pool_id": pool_id,
            "alert_type": alert_type,
        }
        if threshold is not None:
            payload["threshold"] = threshold
        return self.http_client.post("/personal-ai/liquidity/alert", data=payload)

    def list_assistants(self) -> list[dict[str, Any]]:
        """
        List available AI assistants.

        Returns:
            List of available AI assistants

        Raises:
            APIError: If listing fails
        """
        response = self.http_client.get("/personal-ai/assistants")
        return response.get("assistants", [])

    def stream(
        self,
        prompt: str,
        assistant_id: str | None = None,
        context: dict[str, Any] | None = None,
    ) -> Iterator[str]:
        """
        Stream AI response chunks.

        Args:
            prompt: User prompt
            assistant_id: Optional specific assistant to use
            context: Optional context

        Yields:
            Response chunks

        Raises:
            APIError: If streaming fails
        """
        payload = {"prompt": prompt}
        if assistant_id:
            payload["assistant_id"] = assistant_id
        if context:
            payload["context"] = context

        # Use streaming endpoint
        response = self.http_client.post_stream("/personal-ai/stream", data=payload)
        for chunk in response:
            yield chunk

File: src/xai/sdk/typescript/src/clients/ai-client.ts
/**
 * AI Client for XAI SDK
 *
 * Handles all AI-related operations including:
 * - Personal AI assistant operations
 * - Smart contract creation and deployment
 * - Transaction optimization
 * - Blockchain and wallet analysis
 * - Node setup recommendations
 */

import { HTTPClient } from '../utils/http-client';
import { AIError } from '../errors';

/**
 * AI atomic swap parameters
 */
export interface AtomicSwapParams {
  fromCurrency: string;
  toCurrency: string;
  amount: number;
  recipientAddress: string;
}

/**
 * Smart contract creation parameters
 */
export interface CreateContractParams {
  contractType: string;
  parameters: Record<string, unknown>;
  description?: string;
}

/**
 * Smart contract deployment parameters
 */
export interface DeployContractParams {
  contractCode: string;
  constructorArgs?: unknown[];
  gasLimit?: number;
}

/**
 * Transaction optimization parameters
 */
export interface OptimizeTransactionParams {
  transaction: Record<string, unknown>;
  optimizationGoals?: string[];
}

/**
 * Blockchain analysis parameters
 */
export interface AnalyzeBlockchainParams {
  query: string;
  context?: Record<string, unknown>;
}

/**
 * Liquidity alert parameters
 */
export interface LiquidityAlertParams {
  poolId: string;
  alertType: string;
  threshold?: number;
}

/**
 * AI stream parameters
 */
export interface StreamParams {
  prompt: string;
  assistantId?: string;
  context?: Record<string, unknown>;
}

/**
 * AI assistant information
 */
export interface AIAssistant {
  id: string;
  name: string;
  description: string;
  capabilities: string[];
}

/**
 * Client for AI operations on XAI blockchain
 */
export class AIClient {
  constructor(private httpClient: HTTPClient) {}

  /**
   * Execute atomic swap with AI assistance
   *
   * @param params - Swap parameters
   * @returns Swap execution result
   *
   * @example
   * ```typescript
   * const result = await client.ai.atomicSwap({
   *   fromCurrency: 'XAI',
   *   toCurrency: 'BTC',
   *   amount: 100,
   *   recipientAddress: 'bc1q...'
   * });
   * ```
   */
  async atomicSwap(params: AtomicSwapParams): Promise<Record<string, unknown>> {
    try {
      return await this.httpClient.post('/personal-ai/atomic-swap', {
        from_currency: params.fromCurrency,
        to_currency: params.toCurrency,
        amount: params.amount,
        recipient_address: params.recipientAddress,
      });
    } catch (error) {
      throw new AIError(`Atomic swap failed: ${error}`);
    }
  }

  /**
   * Create smart contract with AI assistance
   *
   * @param params - Contract creation parameters
   * @returns Created contract details
   *
   * @example
   * ```typescript
   * const contract = await client.ai.createContract({
   *   contractType: 'token',
   *   parameters: { name: 'MyToken', symbol: 'MTK', supply: 1000000 },
   *   description: 'A custom ERC20 token'
   * });
   * ```
   */
  async createContract(params: CreateContractParams): Promise<Record<string, unknown>> {
    try {
      const payload: Record<string, unknown> = {
        contract_type: params.contractType,
        parameters: params.parameters,
      };
      if (params.description) {
        payload.description = params.description;
      }
      return await this.httpClient.post('/personal-ai/smart-contract/create', payload);
    } catch (error) {
      throw new AIError(`Contract creation failed: ${error}`);
    }
  }

  /**
   * Deploy smart contract with AI optimization
   *
   * @param params - Deployment parameters
   * @returns Deployment result with contract address
   *
   * @example
   * ```typescript
   * const deployment = await client.ai.deployContract({
   *   contractCode: '0x608060...',
   *   constructorArgs: ['MyToken', 'MTK'],
   *   gasLimit: 3000000
   * });
   * console.log('Contract deployed at:', deployment.address);
   * ```
   */
  async deployContract(params: DeployContractParams): Promise<Record<string, unknown>> {
    try {
      const payload: Record<string, unknown> = {
        contract_code: params.contractCode,
      };
      if (params.constructorArgs) {
        payload.constructor_args = params.constructorArgs;
      }
      if (params.gasLimit) {
        payload.gas_limit = params.gasLimit;
      }
      return await this.httpClient.post('/personal-ai/smart-contract/deploy', payload);
    } catch (error) {
      throw new AIError(`Contract deployment failed: ${error}`);
    }
  }

  /**
   * Optimize transaction with AI
   *
   * @param params - Optimization parameters
   * @returns Optimized transaction details
   *
   * @example
   * ```typescript
   * const optimized = await client.ai.optimizeTransaction({
   *   transaction: { to: '0x...', value: '1000' },
   *   optimizationGoals: ['low_fee', 'fast_confirmation']
   * });
   * ```
   */
  async optimizeTransaction(params: OptimizeTransactionParams): Promise<Record<string, unknown>> {
    try {
      const payload: Record<string, unknown> = {
        transaction: params.transaction,
      };
      if (params.optimizationGoals) {
        payload.optimization_goals = params.optimizationGoals;
      }
      return await this.httpClient.post('/personal-ai/transaction/optimize', payload);
    } catch (error) {
      throw new AIError(`Transaction optimization failed: ${error}`);
    }
  }

  /**
   * Analyze blockchain with AI
   *
   * @param params - Analysis parameters
   * @returns Analysis results
   *
   * @example
   * ```typescript
   * const analysis = await client.ai.analyzeBlockchain({
   *   query: 'What are the most active addresses in the last 24 hours?'
   * });
   * ```
   */
  async analyzeBlockchain(params: AnalyzeBlockchainParams): Promise<Record<string, unknown>> {
    try {
      const payload: Record<string, unknown> = { query: params.query };
      if (params.context) {
        payload.context = params.context;
      }
      return await this.httpClient.post('/personal-ai/analyze', payload);
    } catch (error) {
      throw new AIError(`Blockchain analysis failed: ${error}`);
    }
  }

  /**
   * Analyze wallet with AI
   *
   * @param address - Wallet address to analyze
   * @returns Wallet analysis results
   *
   * @example
   * ```typescript
   * const analysis = await client.ai.analyzeWallet('XAI1abc...');
   * console.log('Risk score:', analysis.riskScore);
   * ```
   */
  async analyzeWallet(address: string): Promise<Record<string, unknown>> {
    try {
      return await this.httpClient.post('/personal-ai/wallet/analyze', { address });
    } catch (error) {
      throw new AIError(`Wallet analysis failed: ${error}`);
    }
  }

  /**
   * Get wallet recovery advice from AI
   *
   * @param partialInfo - Partial wallet information
   * @returns Recovery recommendations
   *
   * @example
   * ```typescript
   * const advice = await client.ai.walletRecoveryAdvice({
   *   mnemonicWords: ['apple', 'banana', '...'],
   *   creationDate: '2024-01-01'
   * });
   * ```
   */
  async walletRecoveryAdvice(partialInfo: Record<string, unknown>): Promise<Record<string, unknown>> {
    try {
      return await this.httpClient.post('/personal-ai/wallet/recovery', partialInfo);
    } catch (error) {
      throw new AIError(`Wallet recovery advice failed: ${error}`);
    }
  }

  /**
   * Get node setup recommendations from AI
   *
   * @param hardwareSpecs - Optional hardware specifications
   * @param useCase - Optional use case
   * @returns Setup recommendations
   *
   * @example
   * ```typescript
   * const recommendations = await client.ai.nodeSetupRecommendations(
   *   { ram: '16GB', cpu: '4 cores', storage: '500GB SSD' },
   *   'validator'
   * );
   * ```
   */
  async nodeSetupRecommendations(
    hardwareSpecs?: Record<string, unknown>,
    useCase?: string
  ): Promise<Record<string, unknown>> {
    try {
      const payload: Record<string, unknown> = {};
      if (hardwareSpecs) {
        payload.hardware_specs = hardwareSpecs;
      }
      if (useCase) {
        payload.use_case = useCase;
      }
      return await this.httpClient.post('/personal-ai/node/setup', payload);
    } catch (error) {
      throw new AIError(`Node setup recommendations failed: ${error}`);
    }
  }

  /**
   * Set up liquidity pool alert with AI monitoring
   *
   * @param params - Alert parameters
   * @returns Alert configuration result
   *
   * @example
   * ```typescript
   * const alert = await client.ai.liquidityAlert({
   *   poolId: 'XAI-USDC',
   *   alertType: 'impermanent_loss',
   *   threshold: 5.0
   * });
   * ```
   */
  async liquidityAlert(params: LiquidityAlertParams): Promise<Record<string, unknown>> {
    try {
      const payload: Record<string, unknown> = {
        pool_id: params.poolId,
        alert_type: params.alertType,
      };
      if (params.threshold !== undefined) {
        payload.threshold = params.threshold;
      }
      return await this.httpClient.post('/personal-ai/liquidity/alert', payload);
    } catch (error) {
      throw new AIError(`Liquidity alert setup failed: ${error}`);
    }
  }

  /**
   * List available AI assistants
   *
   * @returns List of available AI assistants
   *
   * @example
   * ```typescript
   * const assistants = await client.ai.listAssistants();
   * assistants.forEach(a => console.log(a.name, a.capabilities));
   * ```
   */
  async listAssistants(): Promise<AIAssistant[]> {
    try {
      const response = await this.httpClient.get<{ assistants: AIAssistant[] }>('/personal-ai/assistants');
      return response.assistants || [];
    } catch (error) {
      throw new AIError(`Failed to list assistants: ${error}`);
    }
  }
}
